design/modular-development.md

```diff
@@ -0,0 +1,288 @@
+# Modular Development with the Azure SDK
+
+**STATUS**: In Progress
+
+Currently, the Azure SDK for JavaScript supports a class-based approach to creating Azure service clients.  This had the advantage of being able to easily identify capabilities of that package.
+
+Pros of this approach includes:
+
+- Pattern follows all other languages
+- Easy discovery of package capabilities through the *Client class.
+- Minimal number of import statements
+
+There are cons to this approach as we consider bringing the JavaScript clients into the browser and service workers where bundle sizes count whether our customers use us or not.  This is especially important as we look at new opportunities such as Web Push which would use a bit of the code from Notification Hubs.
+
+- Large index files to export everything.
+- Not as easily tree shakable to produce smaller bundle sizes.
+- Many methods on the client will never be used, thus taking up valuable space.
+
+## Modular Design
+
+We can create a more ala carte experience where the customer only pulls in the methods they need.  With the Azure SDK for JavaScript, we can create a client previously using the `ServiceClient` pattern:
+
+```typescript
+import { NotificationHubsClient } from "@azure/notification-hubs";
+
+const client = new NotificationHubsClient("<connection-string>", "<hub-name>");
+
+const registrationId = await client.createRegistrationId();
+```
+
+For a more modular design, we have a ServiceClient exposed as an interface with the minimal amount of things to communicate with the backend, and then this is passed into each method.
+
+```typescript
+import { createClientContext, createRegistrationId } from "@azure/notification-hubs/api";
+
+const context = createClientContext("<connection-string>", "<hub-name>");
```

Comment:
How should we handle multi-client libraries? It may be interesting to think about how to future-proof for packages that initially have a single client but later on need to add a new client in a non-breaking way

Reply:
It does seem a little weird that we'd have a bunch of packages that all declare this same-named function. I assume any package following this design would export a function named `createClientContext`. It might be beneficial to defensively name them after the operation group they're a context for, like `createNotificationHubsContext`. In the future, if another client is needed, we could decide whether it needs a new context type or if we can reuse the `NotificationHubsContext`.